diff --git a/app/pktgen.h b/app/pktgen.h
index 4a70a3a..56a6b62 100644
--- a/app/pktgen.h
+++ b/app/pktgen.h
@@ -108,8 +108,8 @@ extern "C" {
 #define MAX_STRING         256
 #define Million            (uint64_t)(1000000ULL)
 
-#define iBitsTotal(_x) (uint64_t)(((_x.ipackets * PKT_OVERHEAD_SIZE) + _x.ibytes) * 8)
-#define oBitsTotal(_x) (uint64_t)(((_x.opackets * PKT_OVERHEAD_SIZE) + _x.obytes) * 8)
+#define iBitsTotal(_x) (uint64_t)((_x.ibytes) * 8)
+#define oBitsTotal(_x) (uint64_t)((_x.obytes) * 8)
 
 #define _do(_exp) \
     do {          \
